### PR TITLE
Add new rule: jsx-use-translation-function

### PR DESCRIPTION
**Pull request by mohsen1**

### CODE REVIEW

**Comment by adidahiya** [comment]
Circle isn't reporting build status; can you look into that please?

File: src/rules/jsxUseTranslationFunctionRule.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "tslint";
+import * as ts from "typescript";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static TRANSLATABLE_ATTRIBUTES = new Set(["placeholder", "title", "alt"]);
```
Comment (mohsen1) [inline src/rules/jsxUseTranslationFunctionRule.ts]: Any other attribute you can think of?

**Comment by mohsen1** [comment]
CI is 👍

File: src/rules/jsxUseTranslationFunctionRule.ts (hunk 1 above)
Comment (saponifi3d) [inline src/rules/jsxUseTranslationFunctionRule.ts]: only one i can think of is href for translated routes.

Maybe make this `Set` be configurable in the tslint config for the `jsx-use-translation-function` options.

**Comment by mohsen1** [comment]
Is this good to merge? 

Note that this rule should probably off by default

File: src/rules/jsxUseTranslationFunctionRule.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public static FAILURE_STRING = "No string literal in JSX. Use a translation function.";
```
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: please be more specific: "String literals are disallowed as JSX attribute values. Use a translation function."

File: src/rules/jsxUseTranslationFunctionRule.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @license
+ * Copyright 2016 Palantir Technologies, Inc.
```
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: nit: 2017

File: src/rules/jsxUseTranslationFunctionRule.ts (hunk 2 above)
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: sorry, strike the part about "attribute values". Just "String literals are disallowed in JSX. Use ..."

File: src/rules/jsxUseTranslationFunctionRule.ts (hunk 3 above)
Comment (mohsen1) [inline src/rules/jsxUseTranslationFunctionRule.ts]: This was generated by the template. Should we make it smarter?

File: src/rules/jsxUseTranslationFunctionRule.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        const walker = new JsxUseTranslationFunctionWalker(sourceFile, this.getOptions());
+        return this.applyWithWalker(walker);
+    }
+}
+
+class JsxUseTranslationFunctionWalker extends Lint.RuleWalker {
+    public visitJsxElement(node: ts.JsxElement) {
+        // TODO: replace this method with visitJsxText for simpler implementation
+        for (const child of node.children) {
+            if (child.kind === ts.SyntaxKind.JsxText && child.getText().trim() !== "") {
```
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: can you use `nodeIsKind<ts.JsxText>(child, ts.SyntaxKind.JsxText)` from `../guards.ts` instead? same for the other checks below

File: src/rules/jsxUseTranslationFunctionRule.ts (hunk 3 above)
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: ha, yeah, should be pretty easy to make the script smarter.

Comment (mohsen1) [inline src/rules/jsxUseTranslationFunctionRule.ts]: https://github.com/palantir/tslint-react/pull/65/commits/12502b62da074d0120b692dcf71e7cb4b0a3fbea

File: src/rules/jsxUseTranslationFunctionRule.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Lint from "tslint";
+import * as ts from "typescript";
+import {nodeIsKind} from "../guards";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static TRANSLATABLE_ATTRIBUTES = new Set(["placeholder", "title", "alt"]);
```
Comment (adidahiya) [inline src/rules/jsxUseTranslationFunctionRule.ts]: Purely syntactic checks will not be enough to fully enforce such a lint rule, but I think this is good enough to be merged in. For example, our `Button` components in https://github.com/palantir/blueprint use a `text: string` attribute which will not be linted by this rule.

**Comment by adidahiya** [comment]
@mohsen1 can you also send a PR for an update to the README to document this rule?

**Pull request merged**